### PR TITLE
 CancellationTokenSource.dispose() should not cancel()

### PR DESCRIPTION
**Pull request by bpasero**

### CODE REVIEW

File: src/vs/base/common/cancellation.ts [hunk 1]
```diff
@@ -93,6 +99,8 @@ export class CancellationTokenSource {
 	}
 
 	dispose(): void {
-		this.cancel();
+		if (this._token instanceof MutableToken) {
```
Comment (bpasero) [inline src/vs/base/common/cancellation.ts]: @jrieken unclear to me what should happen if `dispose` is called without token being there. It almost seems like we should have a new `CancellationToken.Disposed` that is a) not canceled and b) has no-ops for the methods.

Alternatively to keep current semantics we could say that in this case we do `this._token = CancellationToken.Cancelled`

**Comment by onekid1469** [comment]
I hope you don’t take too many people down with you this seems like a good organization and could do great things but with a leader like you only fail

**Comment by eamodio** [comment]
@bpasero Curiously why was this changed? I stumbled upon this when digging through a new GitLens issue that this caused.

**Comment by bpasero** [comment]
@eamodio the motivation is to allow to dispose a cancellation token source without firing the cancellation. Otherwise you would not be able to clean up the resources of the token source (e.g. the event emitter) without also triggering the cancellation. The two are different:
* dispose: free resources
* cancel: indicate that the token is cancelled

**Comment by eamodio** [comment]
@bpasero Thanks, makes sense. Although I wish there was some place to learn about breaking api changes -- at least in the future.

**Comment by bpasero** [comment]
@eamodio every milestone we publish a document called "Release Notes" that shows new features, noteworthy bugfixes but also changes that have an impact for extension authors: https://code.visualstudio.com/updates/v1_22#_extension-authoring

image

**Comment by eamodio** [comment]
@bpasero Wow -- sorry. Somehow I glossed right over that -- and I always read those in entirety. Maybe could just call out breaking changes as such. Anyway, thanks for the help/info.

**Comment by bpasero** [comment]
@eamodio no problemo

**Pull request merged**